### PR TITLE
Add widescreen to Hexen

### PR DESCRIPTION
**Pull request by mikeday0**
Hopefully done? Most of it is copy and paste from Heretic, with the exception of the widescreen asset support. If my approach to that is deemed okay I'll port it over to Heretic so there's parity between the two games.

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
Any plans to finish this? It will be appreciated.

**Comment by mikeday0** [comment]
Absolutely, yes. I hope to knock it out this weekend.

File: src/hexen/f_finale.c [hunk 1]
```diff
@@ -170,8 +170,9 @@ static void TextWrite(void)
     int cx, cy;
     patch_t *w;
 
-    V_CopyScaledBuffer(I_VideoBuffer, W_CacheLumpNum(FinaleLumpNum, PU_CACHE),
-           ORIGWIDTH * ORIGHEIGHT);
+    //V_CopyScaledBuffer(I_VideoBuffer, W_CacheLumpNum(FinaleLumpNum, PU_CACHE),
```
Comment (fabiangreffrath) [inline src/hexen/f_finale.c:173]: Just remove the old code, please.

File: src/hexen/h2def.h [hunk 2]
```diff
@@ -574,6 +574,7 @@ typedef struct player_s
 #define	CF_NOMOMENTUM	4       // not really a cheat, just a debug aid
 
 #define	SBARHEIGHT	(39 << crispy->hires)      // status bar height at bottom of screen
+#define ORIGSBARHEIGHT          39 // [crispy]
```
Comment (fabiangreffrath) [inline src/hexen/h2def.h:577]: If you flip this line with the previous one, you could `#define SBARHEIGHT (ORIGSBARHEIGHT << crispy->hires)`.

File: src/hexen/mn_menu.c [hunk 3]
```diff
@@ -1779,10 +1779,36 @@ void MN_DeactivateMenu(void)
...
-    V_CopyScaledBuffer(I_VideoBuffer,
-           (byte *) W_CacheLumpNum(W_GetNumForName("TITLE") + InfoType,
-                                   PU_CACHE), ORIGWIDTH * ORIGHEIGHT);
+    // V_CopyScaledBuffer(I_VideoBuffer,
+    //        (byte *) W_CacheLumpNum(W_GetNumForName("TITLE") + InfoType,
+    //                                PU_CACHE), ORIGWIDTH * ORIGHEIGHT);
+
+    // [crispy] Refactor to allow for use of V_DrawFullscreenRawOrPatch
+
+    switch (InfoType)
```
Comment (fabiangreffrath) [inline src/hexen/mn_menu.c:1791]: I like that the original code requires no prior knowledge of the names of the next lumps after "TITLE". Do you think we could keep it that way?

File: src/hexen/sb_bar.c [hunk 4]
```diff
@@ -1180,9 +1233,11 @@ void DrawMainBar(void)
          for (j = 0; j <= crispy->hires; j++)
           for (k = 0; k <= crispy->hires; k++)
           {
-            I_VideoBuffer[((i << crispy->hires) + j) * SCREENWIDTH + ((95 << crispy->hires) + k)] = 0;
-            I_VideoBuffer[((i << crispy->hires) + j) * SCREENWIDTH + ((96 << crispy->hires) + k)] = 0;
-            I_VideoBuffer[((i << crispy->hires) + j) * SCREENWIDTH + ((97 << crispy->hires) + k)] = 0;
+            temp = SCREENWIDTH * ((i << crispy->hires) + j);
```
Comment (fabiangreffrath) [inline src/hexen/sb_bar.c:1236]: This change it not really widescreen-related, right? Please add comment.

File: src/hexen/sb_bar.c [hunk 5]
```diff
@@ -1180,9 +1233,11 @@ void DrawMainBar(void)
...
+
```
Comment (fabiangreffrath) [inline src/hexen/sb_bar.c:1237]: Whitespace consistency.

File: src/v_video.h [hunk 6]
```diff
@@ -27,6 +27,8 @@
 // Needed because we are refering to patches.
 #include "v_patch.h"
 
+#include "w_wad.h" // [crispy]
```
Comment (fabiangreffrath) [inline src/v_video.h:30]: For `lumpindex_t`?

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Thank you for this!

Where is the Crispness menu to enable this while playing? 😱

**Comment by mikeday0** [comment]
> Where is the Crispness menu to enable this while playing? scream

Yeah, I know. :( I just controlled it from the config file. Since there's no Crispness menu at all I decided to keep this laser-focused on just the underlying widescreen implementation. I thought that adding the Crispness menu was probably an effort worthy of a separate PR. If that's not out of bounds here I would be happy to add it onto this PR.

**Comment by fabiangreffrath** [comment]
No problem, we can go step by step. Let's finish this widescreen PR first and then a a basic Crispness menu afterwards.

File: src/hexen/mn_menu.c (hunk 3 above)
Comment (mikeday0) [reply src/hexen/mn_menu.c:1791]: The original approach works with the lumps out of hexen.wad but it fails when using replacement graphics from a pwad. (Example included) The original code is dependent on the four lumps being grouped together and having consecutive indices.

[hexen_ws.zip](https://github.com/fabiangreffrath/crispy-doom/files/7882883/hexen_ws.zip)

File: src/hexen/sb_bar.c (hunk 4 above)
Comment (mikeday0) [reply src/hexen/sb_bar.c]: I went ahead and removed the use of temp.

File: src/v_video.h (hunk 6 above)
Comment (mikeday0) [reply src/v_video.h]: Comment added.

**Comment by fabiangreffrath** [comment]
Thank you!

**Pull request merged**